server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java

```diff
@@ -287,6 +281,49 @@ public void testResetDuringInFlightRequest() throws Exception {
         assertBusy(() -> assertThat(requestCounter.get(), equalTo(2)));
     }
 
+    /**
+     * This test verifies that we correctly cancel a Monitoring instance when we only just started the instance.
+     * In other words, we're verifying the behaviour when the monitor is cancelled <i>before</i> the <code>threadPool.schedule(...)</code>
+     * invocation returns.
+     */
+    public void test() throws Exception {
+        ClusterState initialState = ClusterStateCreationUtils.state(node, node, node, new DiscoveryNode[] { node, frozenNode })
+            .copyAndUpdate(b -> b.putCustom(HealthMetadata.TYPE, healthMetadata));
+        when(clusterService.state()).thenReturn(initialState);
+
+        var cancelableRef = new AtomicReference<LocalHealthMonitor.Monitoring>();
+        var counter = new AtomicInteger();
+        doAnswer(invocation -> {
+            var cancellable = ((Scheduler.ScheduledCancellable) invocation.callRealMethod());
+            var currentCount = counter.incrementAndGet();
+            if (currentCount == 1) {
+                LocalHealthMonitor.Monitoring monitoring = invocation.getArgument(0);
+                cancelableRef.set(monitoring);
+                assertBusy(() -> assertTrue(monitoring.isCancelled()));
+            }
+            return cancellable;
+        }).when(threadPool).schedule(any(), any(), any());
+
+        doAnswer(invocation -> {
+            ActionListener<AcknowledgedResponse> listener = invocation.getArgument(2);
+            listener.onResponse(null);
+            return null;
+        }).when(client).execute(any(), any(), any());
+
+        localHealthMonitor.setMonitorInterval(TimeValue.timeValueMillis(10));
+        // Run in a new thread because the `threadPool.schedule(...)` invocation will hang (as defined above).
+        new Thread(() -> localHealthMonitor.clusterChanged(new ClusterChangedEvent("start-up", initialState, ClusterState.EMPTY_STATE)))
+            .start();
+        assertBusy(() -> assertThat(counter.get(), greaterThanOrEqualTo(2)));
```

Comment:
I don't fully get this assertion. I think I am missing something here.

If I understand correctly on line 315-316 we set the write conditions to start the monitoring. The monitoring will block on the first time it's trying to start because `currentCount` will be 1. Where is the second call to the `threadpool.schedule` coming from?

Reply:
That's coming from `Monitoring#scheduleNextRunIfNecessary`:https://github.com/elastic/elasticsearch/blob/f35caa2554b7e7e4608dc69bf1f8380a38bf662c/server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java#L398-L398

And for clarification, the monitoring doesn't exactly "block". When we call `threadPool.schedule(...)` here https://github.com/elastic/elasticsearch/blob/f35caa2554b7e7e4608dc69bf1f8380a38bf662c/server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java#L282-L282
We do actually execute `Monitoring#run`, as per https://github.com/elastic/elasticsearch/blob/f35caa2554b7e7e4608dc69bf1f8380a38bf662c/server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java#L297-L297
but we don't _return_ yet (i.e. we block before we return). That means the monitoring will be started, but `scheduledRun` will not be assigned yet because `threadPool.schedule(...)` hasn't returned yet.

Does that make things more clear? Do you have suggestions for comments/restructuring that would make this easier to read?

Reply:
Hm, I see. So, let's walk us through this:

1) We start the monitoring, which will schedule the run but it will not return so the `scheduledRun` will not be assigned.
https://github.com/elastic/elasticsearch/blob/f35caa2554b7e7e4608dc69bf1f8380a38bf662c/server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java#L281-L283

2) The monitoring will `run` and will try to schedule the next run. This will succeed because we only apply the disruption on the first execution.
https://github.com/elastic/elasticsearch/blob/f35caa2554b7e7e4608dc69bf1f8380a38bf662c/server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java#L326

3) When we cancel the Monitoring:
https://github.com/elastic/elasticsearch/blob/f35caa2554b7e7e4608dc69bf1f8380a38bf662c/server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java#L317-L318
At this moment the `scheduledRun` actually has been set with the second scheduled job, based on my understanding. So, we are not actually cancelling the first run.

So, I think this test is testing something else. What if we changed it slightly:

1) We actually block before the execution, so we do not return and we do not run until we are sure that the cancel signal has been sent.
2) When the cancel signal has been sent, we unblock, we run and we verify that there shouldn't be any request to the health node since this has been already cancelled.

I think the above is capturing an early cancellation better. What do you think?

Reply:
As discussed with Andrei, I'm removing this test :). Thanks for the feedback anyway!